toolkit/tools/pkg/imagecustomizerlib/customizeuki.go

```diff
@@ -0,0 +1,538 @@
+
+	var err error
+
+	loopback, err := safeloopback.NewLoopback(buildImageFile)
+	if err != nil {
+		return fmt.Errorf("failed to connect to image file to provision UKI:\n%w", err)
+	}
+	defer loopback.Close()
+
+	diskPartitions, err := diskutils.GetDiskPartitions(loopback.DevicePath())
+	if err != nil {
+		return err
+	}
+
+	systemBootPartition, err := findSystemBootPartition(diskPartitions)
+	if err != nil {
+		return err
+	}
+	bootPartition, err := findBootPartitionFromEsp(systemBootPartition, diskPartitions, buildDir)
+	if err != nil {
+		return err
+	}
+
+	systemBootPartitionTmpDir := filepath.Join(buildDir, tmpEspParitionDirName)
+	systemBootPartitionMount, err := safemount.NewMount(systemBootPartition.Path, systemBootPartitionTmpDir, systemBootPartition.FileSystemType, 0, "", true)
+	if err != nil {
+		return fmt.Errorf("failed to mount esp partition (%s):\n%w", bootPartition.Path, err)
+	}
+	defer systemBootPartitionMount.Close()
+
+	bootPartitionTmpDir := filepath.Join(buildDir, tmpParitionDirName)
+	bootPartitionMount, err := safemount.NewMount(bootPartition.Path, bootPartitionTmpDir, bootPartition.FileSystemType, 0, "", true)
+	if err != nil {
+		return fmt.Errorf("failed to mount partition (%s):\n%w", bootPartition.Path, err)
+	}
+	defer bootPartitionMount.Close()
+
+	osSubreleaseFullPath := filepath.Join(buildDir, UkiBuildDir, "os-release")
+	stubPath := filepath.Join(buildDir, UkiBuildDir, "linuxx64.efi.stub")
+	cmdlineFilePath := filepath.Join(buildDir, UkiBuildDir, KernelCmdlineArgs+".json")
+
+	// Get mapped kernels and initramfs.
+	kernelToInitramfs, err := getKernelToInitramfsMap(bootPartitionTmpDir, uki.Kernels)
+	if err != nil {
+		return err
+	}
+
+	for kernel, initramfs := range kernelToInitramfs {
+		err := buildUki(kernel, initramfs, cmdlineFilePath, osSubreleaseFullPath, stubPath, buildDir,
+			systemBootPartitionTmpDir,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to build UKI for kernel (%s):\n%w", kernel, err)
+		}
+	}
+
+	err = cleanupUkiBuildDir(buildDir)
+	if err != nil {
+		return fmt.Errorf("Error during cleanup UKI build dir:\n%w", err)
+	}
+
+	err = cleanupBootPartition(bootPartitionTmpDir)
+	if err != nil {
+		return fmt.Errorf("failed to clean up boot partition:\n%w", err)
+	}
+
+	err = systemBootPartitionMount.CleanClose()
+	if err != nil {
+		return err
+	}
+
+	err = bootPartitionMount.CleanClose()
+	if err != nil {
+		return err
+	}
+
+	err = loopback.CleanClose()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Note: This function will be optimized by leveraging the internal functions
+// under grubcfgutils.go when implementing bootloader customization.
+func extractKernelToArgsFromGrub(grubCfgPath string) (map[string]string, error) {
+	grubCfgContent, err := file.Read(grubCfgPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read grub.cfg file at (%s):\n%w", grubCfgPath, err)
+	}
+
+	linuxLineRegex := regexp.MustCompile(`^linux\s+(/vmlinuz-[^\s]+)\s+(.*)`)
```

Comment:
You should use `findGrubCommandAll` here, instead of manually parsing the grub.cfg file.